### DISCUSSION TITLE
Using geodetic coordinates (WGS84) in Bevy

### DISCUSSION THREAD

**Discussion by nehalem501** [bevyengine/bevy#9783] [Q&A]
Hello,

As a hobby project, I would like to try developing some kind of train driving simulator with bevy.

All the position of the tracks and scenery (buildings, vegetation, ...) would be in geodetic coordinates using WGS84. Also, this is the way the already available data exists (OSM, train companies open data, ...). But to display it with the GPU we need it to work in xyz coordinates.

I know that there are formulas for this, but how could I do it efficiently? Also, I don't want to have to recompute everything for each frame, but how can it work if the camera moves every frame ? Because generally this type of conversion has less and less precision as you move further from the origin you chose, and in this type of game you're supposed to move in dozens of km. The place where precision matters is actually close to the camera (wherever this camera is).

Flight simulators have a similar problem to solve, but I couldn't really figure out how it works when I've tried to look at the FlightGear code (open-source flight simulator).

**Comment by bushrat011899** [comment]
I've done a little GIS work outside of Bevy. In my opinion, you would want to pre-process your assets out of WGS84 and into Bevy's coordinate system, either at startup or compile time. Of course, that depends on the types of assets you're working with. If you're streaming in GIS information from something like ArcGIS during gameplay, then obviously this would have to happen at runtime.

As for how to do that mapping most efficiently, I would keep it simple and define a `TransformWGS84` type and the required `From` / `Into` to get into a Bevy `Transform`. If you're working in a small enough area, you might only need to transform the position. But if you want to handle a larger area, you will need to also account for the rotation.

**Reply by nehalem501** [reply]
A rail network is not just a single line, It is a network of lines branching off and merging many kilometers further. So it seems like I will have problems over longer distances because of earth's curvature if I pre-process the assets to convert everything to cartesian xyz coordinates, wouldn't I?
I would like my game to work with a single huge map spanning whole Europe, in a kind of open-world fashion, (This is a map of the European rail network [https://www.openrailwaymap.org/](https://www.openrailwaymap.org/))

This is why I talked about flight simulators, because it seems like a very similar problem when someone flies a plane accross continents in these.

And yes, to have something usable without spending too much time to design the maps, I was thinking about using all the OpenStreetMap data available and stream it at runtime for the time being.

**Reply by bushrat011899** [reply]
You won't have any accuracy issues pre-processing (or on the fly either). The reason we have issues converting out of WGS84 into XYZ is because a normal person with a 2D map can't represent that data losslessly. But you still can in a 3D space, if you allow it to 

For example, you can't assume gravity is just "down": it's towards a fixed point, the centre of the earth.

Basically, you're going to render a giant sphere, and place the camera at a to-scale height above that sphere to replicate a WGS84 version of earth. You'll just need to fix a few things, such as ensuring the camera's up makes sense for your position.

**Comment by nehalem501** [comment]
Seems like @alice-i-cecile have mentionned GIS integration with bevy here https://github.com/bevyengine/bevy/discussions/1678#discussioncomment-496654

**Reply by alice-i-cecile** [reply]
Yep, this is something that you may want to reach out to [Foresight](https://www.foresightmining.com/) about; particularly @aevyrie :) They use Bevy commercially, and working with geographical data is essential when you're at the scale of "real life mines".
